src/bandersnatch_storage_plugins/s3.py

```diff
@@ -0,0 +1,368 @@
+        assert isinstance(file2_contents, bytes)
+        file1_hash = hashlib.sha256(file1_contents).hexdigest()
+        file2_hash = hashlib.sha256(file2_contents).hexdigest()
+        return file1_hash == file2_hash
+
+    def copy_file(self, source: PATH_TYPES, dest: PATH_TYPES) -> None:
+        if not self.exists(source):
+            raise FileNotFoundError(source)
+        if isinstance(source, self.PATH_BACKEND):
+            client = self.resource.meta.client
+            client.copy_object(Key=dest.key, CopySource=source.key, Bucket=dest.bucket)
+        else:
+            with source.open(mode="rb") as fh:
+                with dest.open(mode="wb") as fh2:
+                    fh2.write(fh.read())
+        return
+
+    def write_file(self,
+                   path: PATH_TYPES,
+                   contents: Union[str, bytes],
+                   encoding: Optional[str] = None
+                   ) -> None:
+        if not isinstance(path, self.PATH_BACKEND):
+            path = self.PATH_BACKEND(path)
+        if isinstance(contents, str):
+            with path.open(mode="w", encoding=encoding) as fp:
+                fp.write(contents)
+        elif isinstance(contents, bytes):
+            with path.open(mode="wb") as fp:
+                fp.write(contents)
+        return
+
+    def open_file(
+            self, path: PATH_TYPES, text: bool = True, encoding: str = "utf-8"
+    ) -> TextIO:
+        if not isinstance(path, self.PATH_BACKEND):
+            path = self.PATH_BACKEND(path)
+        mode = "r" if text else "rb"
+        kwargs: Dict[str, str] = {}
+        if text:
+            kwargs["encoding"] = encoding
+        return path.open(mode=mode, **kwargs)
+
+    def read_file(
+            self,
+            path: PATH_TYPES,
+            text: bool = True,
+            encoding: str = "utf-8",
+            errors: Optional[str] = None,
+    ) -> Union[str, bytes]:
+        """Return the contents of the requested file, either a a bytestring or a unicode
+        string depending on whether **text** is True"""
+
+        fh = self.open_file(path, text=text, encoding=encoding)
+        contents: Union[str, bytes] = fh.read()
+        return contents
+
+    def delete_file(self, path: PATH_TYPES, dry_run: bool = False) -> int:
+        """Delete the provided path, recursively if necessary."""
+        if not isinstance(path, self.PATH_BACKEND):
+            path = self.PATH_BACKEND(path)
+        log_prefix = "[DRY RUN] " if dry_run else ""
+        logger.info(f"{log_prefix}Removing file: {path!s}")
+        if not dry_run:
+            path.unlink()
+        return 0
+
+    def mkdir(
+            self, path: PATH_TYPES, exist_ok: bool = False, parents: bool = False
+    ) -> None:
+        """
+        Create the provided directory
+
+        This operation is a no-op on swift.
+        """
+        logger.warning(
+            f"Creating directory in object storage: {path} with .swiftkeep file"
```

Comment:
Should we use a variable here?
```suggestion
            f"Creating directory in object storage: {path} with .s3keep file"
```

Reply:
yes , great idea 🤦‍